newrelic-agent/src/test/java/com/newrelic/agent/transport/DataSenderImplConnectCycleTest.java

```diff
@@ -64,6 +64,35 @@ public void testFullPreconnectConnectCycleNoSecurity() throws Exception {
         assertEquals("value", result.get("other"));
     }
 
+    @Test
+    public void testReconnectCycleNoSecurity() throws Exception {
+        MockitoAnnotations.initMocks(this);
+
+        final Map<String, Object> settings = new HashMap<>();
+        settings.put(AgentConfigImpl.APP_NAME, "Unit Test");
+        settings.put(AgentConfigImpl.HOST, "no-collector.example.com");
+
+        final MockServiceManager serviceManager = new MockServiceManager(
+                ConfigServiceFactory.createConfigServiceUsingSettings(settings)
+        );
+        serviceManager.setStatsService(mockStatsService);
+
+        HttpClientWrapper wrapper = new ConnectCycleNoSecuritySuccessClientWrapper();
+
+        DataSenderImpl target = new DataSenderImpl(serviceManager.getConfigService().getDefaultAgentConfig(), wrapper, null, Agent.LOG,
+                ServiceFactory.getConfigService());
+        target.setAgentRunId("agent run id");
+
+        Map<String, Object> startupOptions = new HashMap<>();
+        startupOptions.put("test-sentinel", "test-value");
+        Map<String, Object> result = target.connect(startupOptions);
+        assertEquals("my-run-id", target.getAgentRunId());
+        assertEquals("value", result.get("other"));
+
+        // This will assert that preconnect was called with the original host, not the redirect host.
+        target.connect(startupOptions);
```

Comment:
it is unclear to me how this would do any assertions around preconnect.  Does this test actually check what this comment says it is checking?  If so, how does it do that?

Reply:
These tests use the `DataSenderImpl` with a customized `HttpClientWrapper`. `HttpClientWrapper` is the interface that masks almost all the Apache HTTP client code. On line 80, this test applies the [`ConnectCycleNoSecuritySuccessClientWrapper`](https://github.com/newrelic/newrelic-java-agent/blob/preconnect-original/newrelic-agent/src/test/java/com/newrelic/agent/transport/ConnectCycleNoSecuritySuccessClientWrapper.java) wrapper. When a `preconnect` URL comes through that wrapper, it verifies that the host is `no-collector.example.com` and sends a redirect host of `new_host.example.com`.

Without the code change, this code would fail because the second preconnect would use the host `new_host.example.com`.

Reply:
Ok, thanks for clarifying...because that's pretty complicated for testing.  At least it is tested.  Thanks!